java/ql/lib/semmle/code/java/CompilationUnit.qll

```diff
@@ -31,5 +31,38 @@ class CompilationUnit extends Element, File {
    */
   Module getModule() { cumodule(this, result) }
 
+  /**
+   * Gets a type which is available in the top-level scope of this compilation unit.
+   * This can be a type:
+   * - declared in this compilation unit as top-level type
+   * - imported with an `import` declaration
+   * - declared in the same package as this compilation unit
+   * - declared in the package `java.lang`
+   *
+   * This predicate not consider "shadowing", it can have types as result whose simple name is
+   * shadowed by another type in scope.
+   */
+  ClassOrInterface getATypeInScope() {
+    // See "Shadowing", https://docs.oracle.com/javase/specs/jls/se17/html/jls-6.html#jls-6.4.1
+    // Currently shadowing is not considered
+    result.(TopLevelType).getCompilationUnit() = this
```

Comment:
Redundant w.r.t. the same-package test

Reply:
You are right. Currently this structure matches (roughly?) the precedence order of shadowing, but I could drop this nonetheless if you want.